File: tests/bench/io_random_pread.c
/**
 * io_random_pread.c — I/O 基线：随机 offset pread（C -O2 对照，PERF-002）
 * 路径须与 run-perf-io.sh 生成的 tests/bench/.io_mmap_bench_tmp 一致。
 */
#include <fcntl.h>
#include <stdint.h>
#include <stdio.h>
#include <unistd.h>

enum { CHUNK = 4096, PAGES = 4096, ROUNDS = 1024 };

int main(void) {
  const char *path = "tests/bench/.io_mmap_bench_tmp";
  int fd = open(path, O_RDONLY);
  if (fd < 0) {
    perror("open");
    return 1;
  }
  uint8_t buf[CHUNK];
  int32_t sum = 0;
  uint32_t seed = 12345u;
  for (int r = 0; r < ROUNDS; r++) {
    seed = seed * 1103515245u + 12345u;
    uint32_t page = seed % PAGES;
    off_t off = (off_t)page * CHUNK;
    ssize_t nr = pread(fd, buf, CHUNK, off);
    if (nr != CHUNK) {
      perror("pread");
      close(fd);
      return 2;
    }
    for (int bi = 0; bi < CHUNK; bi++) {
      sum += (int32_t)buf[bi];
    }
  }
  if (close(fd) != 0) {
    perror("close");
    return 3;
  }
  return sum & 255;
}

File: tests/bench/io_random_pread.zig
// io_random_pread.zig — I/O 基线：随机 offset pread（Zig -O2 对照，PERF-002）
const std = @import("std");

const SEG: usize = 4096;
const PAGES: u32 = 4096;
const ROUNDS: i32 = 1024;

pub fn main() !void {
    const path = "tests/bench/.io_mmap_bench_tmp";
    const file = try std.fs.cwd().openFile(path, .{});
    defer file.close();
    var buf: [SEG]u8 = undefined;
    var sum: i32 = 0;
    var seed: u32 = 12345;
    var r: i32 = 0;
    while (r < ROUNDS) : (r += 1) {
        seed = seed *% 1103515245 +% 12345;
        const page: u32 = seed % PAGES;
        const off: u64 = @as(u64, page) * SEG;
        const nr = try file.preadAll(&buf, off);
        if (nr != SEG) std.process.exit(2);
        for (buf) |b| sum +%= @as(i32, @intCast(b));
    }
    std.process.exit(@intCast(sum & 255));
}
